enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ThreadedTransaction.java

```diff
@@ -102,22 +104,34 @@ public Throwable apply( S subject )
                                 if ( !startEarly )
                                 {
                                     latch.startAndWaitForAllToStart();
+                                    state = LatchState.STARTED;
                                 }
                             }
                             finally
                             {
-                                if ( !startEarly )
+                                if ( !startEarly && state == LatchState.NOT_STARTED )
                                 {
                                     latch.start();
+                                    state = LatchState.STARTED;
                                 }
                                 latch.finishAndWaitForAllToFinish();
+                                state = LatchState.FINISHED;
                             }
                             result.close();
                             tx.success();
                             return null;
                         }
                         catch (Throwable t)
                         {
+                            if ( state == LatchState.NOT_STARTED )
+                            {
+                                latch.start();
+                                state = LatchState.STARTED;
+                            }
+                            if ( state == LatchState.STARTED )
+                            {
+                                latch.finish();
+                            }
```

Comment:
Eh, how is this related to rate limiting? Confused. And also I don't really understand when this code is supposed to kick in - the inner finally block ends with a `state = LatchState.FINISHED` so it should always be finished?